### ISSUE TITLE
Interruption of Semaphore Waits

### ISSUE THREAD

**Issue by rzeigler** [rzeigler/waveguide#7]
I can see situations where it would be useful to interrupt Fibers waiting on a Queue without needing a reference to the fibers themselves to facilitate shutdown semantics. Deferred already supports pushing an interrupt into another fiber, however, in the case of a blocking queue it is conceivable the waiters are waiting on the semaphore acquire rather than the deferred wait. 
We should leverage the interruptibility propagation from Deferred to interrupt all Semaphore waiters (shutdown a semaphore) which will also facilitate similar shutdown semantics of a Queue.
